flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/LogicalFile.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.filemerging;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.util.StringBasedID;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.UUID;
+
+import static org.apache.flink.runtime.checkpoint.filemerging.FileMergingSnapshotManager.SubtaskKey;
+
+/** An abstraction of logical files in file-merging checkpoints. */
+public class LogicalFile {
+
+    /** ID for {@link LogicalFile}. It should be unique for each file. */
+    public static class LogicalFileId extends StringBasedID {
+
+        public LogicalFileId(String keyString) {
+            super(keyString);
+        }
+
+        public Path getFilePath() {
+            return new Path(getKeyString());
+        }
+
+        public static LogicalFileId generateRandomId() {
+            return new LogicalFileId(UUID.randomUUID().toString());
+        }
+    }
+
+    /** ID for this file. */
+    LogicalFileId fileId;
+
+    private long lastCheckpointId = -1L;
+
+    boolean removed = false;
+
+    /** The physical file where this logical file is stored. This should be null. */
+    @Nonnull private final PhysicalFile physicalFile;
+
+    @Nonnull private final SubtaskKey subtaskKey;
+
+    public LogicalFile(
+            LogicalFileId fileId,
+            @Nonnull PhysicalFile physicalFile,
+            @Nonnull SubtaskKey subtaskKey) {
+        this.fileId = fileId;
+        this.physicalFile = physicalFile;
+        this.subtaskKey = subtaskKey;
+        physicalFile.incRefCount();
+    }
+
+    public LogicalFileId getFileId() {
+        return fileId;
+    }
+
+    public void discardWithCheckpointId(long checkpointId) throws IOException {
+        if (!removed && checkpointId >= lastCheckpointId) {
+            physicalFile.decRefCount();
+            removed = true;
+        }
+    }
+
+    public void advanceLastCheckpointId(long checkpointId) {
```

Comment:
I added JavaDocs to explain this method. In short, the value `lastUsedCheckpointId` acts pretty much like the `lastUsedCheckpointID` in `SharedStateEntry`. It is a watermark mechanism to record the last checkpoint that uses this logical file, to determine whether to delete a logical file when a cp is subsumed or aborted. For those two questions:
1. No, a logical file is a shared state file before merging, it may shared across checkpoints.
2. When a file is used/reused by a checkpoint, this method is called.